### PR TITLE
hal/ia32: apply new syspage

### PR DESCRIPTION
**Pull request by HBuczynski**
## Description

 - fix initialization of `_interrupts_multilock` variable
 - apply new syspage approach
 - update console

**NOTE:** console-vga doesn't work correctly. `lib_printf()` sets console color using ASCI what is incomprehensible to vga.
Changes will be introduced in a separate PR.
Furthermore, **console-serial** has hard-coded instance of base address. This code will be changed in the next PR as well.
 
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (lia32-generic).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (https://github.com/phoenix-rtos/plo/pull/157).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: hal/ia32/halsyspage.h [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * %LICENSE%
+ */
+
+#ifndef _HAL_SYSPAGE_H_
+#define _HAL_SYSPAGE_H_
+
+#include "config.h"
+
+extern hal_syspage_t *syspage;
```
Comment (agkaminski) [inline hal/ia32/halsyspage.h:21]: Hmm, we already have this symbol, don't we? So this is it's redefinition?

File: hal/ia32/pmap.c [hunk 2]
```diff
@@ -235,40 +226,48 @@ int pmap_getPage(page_t *page, addr_t *addr)
...
-	if (k == syspage->mmsize)
-		return -EINVAL;
-
-	/* Page is allocated by BIOS */
-	if (!(page->flags & PAGE_FREE)) {
-		page->flags |= PAGE_OWNER_BOOT;
-		return EOK;
-	}
-
+#if 0
```
Comment (agkaminski) [inline hal/ia32/pmap.c:270]: You added #if 0, so maybe remove comment that comments-out the code inside?

File: include/arch/syspage-imxrt.h [hunk 3]
```diff
@@ -20,13 +20,13 @@
...
+		unsigned int type;
+		unsigned int allocCnt;
 		struct {
-			u32 rbar;
-			u32 rasr;
+			unsigned int rbar;
+			unsigned int rasr;
 		} table[16] __attribute__((aligned(8)));
-		u32 map[16]; /* ((u32)-1) = map is not assigned */
+		unsigned int map[16]; /* ((unsigned int)-1) = map is not assigned */
```
Comment (agkaminski) [inline include/arch/syspage-imxrt.h:29]: Unrelated change and I'm not sure if we want to change it that way

File: hal/ia32/halsyspage.h [hunk 4]
```diff
@@ -0,0 +1,23 @@
+/*
```
Comment (agkaminski) [inline hal/ia32/halsyspage.h:1]: Do we really need this file?

File: hal/ia32/pmap.c [hunk 5]
```diff
@@ -278,23 +277,23 @@ int pmap_getPage(page_t *page, addr_t *addr)
 	}*/
 
 	hal_spinlockClear(&pmap_common.lock, &sc);
+#endif
 
-	if ((page->addr >= (u32)syspage - VADDR_KERNEL) && (page->addr < (u32)syspage - VADDR_KERNEL + SIZE_PAGE)) {
+	/* Calculate syspage size */
+	sz = *(ptr_t *)((ptr_t)syspage + sizeof(hal_syspage_t));
```
Comment (agkaminski) [inline hal/ia32/pmap.c:283]: Why not `syspage->size`?

File: hal/ia32/pmap.c (hunk 2 above)
Comment (HBuczynski) [reply hal/ia32/pmap.c:270]: What do you think about removing it permanently? 
This piece of code describes `ptable` which has a new approach in `pmap_getPage` function:

https://github.com/phoenix-rtos/phoenix-rtos-kernel/blob/3b85faa3db0cb156f678c30dc1d430cc9f0a8f00/hal/ia32/pmap.c#L295-L298

File: include/arch/syspage-imxrt.h (hunk 3 above)
Comment (HBuczynski) [reply include/arch/syspage-imxrt.h:29]: You have right that it is not related to the context of this PR but I have added appropriate comment in commit message. 
`u32` and `u8` are described in `hal`. This code is shared between plo and kernel so it will be better to not use types which are defined in both modules separately.

File: hal/ia32/pmap.c (hunk 5 above)
Comment (HBuczynski) [reply hal/ia32/pmap.c:283]: `syspage` is a type of `hal_syspage_t *`. `size` is a part of generic syspage structure `syspage_t.`

Comment (agkaminski) [reply hal/ia32/pmap.c:283]: It would be better to handle main `syspage` symbol here. You include syspage.h anyway, so why not to access generic syspage by `syspage->field` and `hal_syspage_t` by `syspage->hs->field`?

File: hal/ia32/halsyspage.h (hunk 4 above)
Comment (HBuczynski) [reply hal/ia32/halsyspage.h:1]: The answer is above.

File: hal/ia32/pmap.c (hunk 5 above)
Comment (HBuczynski) [reply hal/ia32/pmap.c:283]: The main assumption was that hal has an access only to `hal_syspage_t` part of the syspage structure. 
`pmap_getPage` uses fields from generic part of syspage but it can be moved outside of hal, on each of the platform with MMU code looks similar.

Your proposition will solve the problem at the moment but I think that we should follow the rule that hal gets from the outside world only the minimum.

File: hal/ia32/cpu.c [hunk 6]
```diff
@@ -445,7 +442,7 @@ static void _hal_cpuInitCores(void)
 	cpu.ncpus = 0;
 	_cpu_initCore();
 
-	*(u32 *)((void *)syspage->stack + VADDR_KERNEL - 4) = 0;
+	*(u32 *)((void *)syspage->hs.stack + VADDR_KERNEL - 4) = 0;
```
Comment (agkaminski) [inline hal/ia32/cpu.c:445]: void * arithmetic

File: hal/ia32/pmap.c [hunk 7]
```diff
@@ -417,38 +398,39 @@ int pmap_segment(unsigned int i, void **vaddr, size_t *size, int *prot, void **t
...
+		if (map->start < pmap_common.minAddr)
+			pmap_common.minAddr = map->start;
+
+		if (map->end > pmap_common.maxAddr)
+			pmap_common.maxAddr = map->end;
+	} while ((map = map->next) != syspage->maps);
 
 	/* Initialize kernel page table - remove first 4 MB mapping */
-	pmap->pdir = VADDR_KERNEL + (void *)syspage->pdir;
+	pmap->pdir = VADDR_KERNEL + (void *)syspage->hs.pdir;
```
Comment (agkaminski) [inline hal/ia32/pmap.c:423]: void * arithmetic

File: include/arch/syspage-imxrt.h (hunk 3 above)
Comment (agkaminski) [reply include/arch/syspage-imxrt.h:29]: Ok for now, but imho this types might be/should be shared between plo and kernel in the future

File: hal/ia32/pmap.c (hunk 7 above)
Comment (HBuczynski) [reply hal/ia32/pmap.c]: Fixed

**Review by agkaminski — APPROVED** [review]
LGTM, let's wait for @Maxez

**Review by Maxez — APPROVED** [review]
One thing to change - we've agreed that `syspage` variable points to the whole syspage (not just the hal part) so we should change its definition in hal.c (hal_syspage_t * needs to be changed to syspage_t *).

**Comment by HBuczynski** [comment]
> One thing to change - we've agreed that `syspage` variable should point to the whole syspage (not just the hal part) so we should change its definition in hal.c (`hal_syspage_t *` needs to be changed to `syspage_t *`).

Thanks, done.
